feat: replacing concrete client keeper struct with an interface

SebastianElvis opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

This PR replaces the concrete `ClientKeeper` struct with an interface. This allows IBC-enabled applications to add extra functionalities atop the client keeper (without forking IBC-Go), including

- adding hooks to these keepers, so that other modules can learn internal states of IBC
- defining and emitting new events upon certain state changes in the IBC module
- implementing new queries and messages for these internal keepers

closes: #2880 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

crodriguezvega commented:
Thanks for this contribution, @SebastianElvis! We will discuss and review in the coming weeks. 

Do you already have a concrete use case to use this? Do you have any urgency to have this released soon?

SebastianElvis commented:
Thanks for the response!

> Do you already have a concrete use case to use this?

Yep, we indeed have a concrete use case -- adding hooks over `ClientKeeper` for checkpointing light client headers.

Our project [Babylon](https://babylonchain.io/) aims at building a blockchain Babylon that checkpoints other PoS chains to Bitcoin. The checkpointing process from PoS chains to Babylon leverages the IBC light client, where Babylon receives PoS chain's headers via IBC. In order to do the checkpointing, Babylon has to use hooks to get notification at the application layer upon receiving headers from IBC.

At the moment, we are maintaining [a fork of IBC-Go](https://github.com/babylonchain/ibc-go) that implements a hook [AfterHeaderWithValidCommit](https://github.com/babylonchain/ibc-go/blob/babylon-v5.0.0/modules/core/02-client/keeper/extended_keeper_hooks.go#L33-L37). The hook is invoked upon a header with a valid quorum certificate from IBC. The most code of Babylon is still WIP and we hope to open-source everything as soon as possible.

> Do you have any urgency to have this released soon?

Yeah we definitely hope to have this feature in the official IBC-Go implementation. Nevertheless, we respect your priorities. Before having this feature in IBC-Go we can still use our fork :) Thanks for asking.

colin-axner commented on testing/simapp/app.go:
```diff
@@ -893,13 +894,18 @@ func (app *SimApp) setupUpgradeHandlers() {
+
 	app.UpgradeKeeper.SetUpgradeHandler(
 		v7.UpgradeName,
 		v7.CreateUpgradeHandler(
```
same with this one

colin-axner commented on modules/core/02-client/exported/exported.go:
```diff
@@ -0,0 +1,68 @@
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+)
+
+// ClientKeeper defines the expected interface of the light client used for IBC
```
Maybe we could add some short documentation explaining why this interface exists? That is to extend the ClientKeeper implementation (the ibc module client keeper implementation should not have its logic modified, only extended)

colin-axner replied:
Maybe we could add additional documentation [here](https://github.com/cosmos/ibc-go/blob/main/docs/ibc/integration.md), via a section "Extending the client keeper"

colin-axner reviewed (commented):
Thanks @SebastianElvis! The changes look good! Left some minor comments, let me know if you have any questions

SebastianElvis commented:
Thanks @colin-axner for the feedback, which is really helpful! I have rebased the PR to the latest `main` branch, and fixed the comments. Specifically, this PR 

- changes the signatures of `NewMigrator`, `PruneExpiredConsensusStates`, and `SetUpgradeHandler`
- removes unnecessary type assertions given the updated function signatures
- adds documentations for `clientexported.ClientKeeper` and integration

Please feel free to have a look again.

chatton commented:
Hi @SebastianElvis 

Thanks for all the work you've been doing here, looking at the code, I am a little concerned about the size of the interface that ends up being required here, and a lot of these interface functions are actually implementation details (un/marshalling functions, client id generation etc. )

I was investigating an [alternative approach](https://github.com/cosmos/ibc-go/pull/3214) which would allow custom logic to be executed at different client events (creation/upgrading/updating) through the use of hooks. I believe this should be enough to enable any sort of custom logic that is required while also making sure that the client keeper logic is guaranteed to execute as normal.

Please let me know what you think of this approach and if it is enough to implement the logic you want to add.

cc @colin-axner

SebastianElvis commented:
Hi @chatton , thanks for the feedback!

> Thanks for all the work you've been doing here, looking at the code, I am a little concerned about the size of the interface that ends up being required here, and a lot of these interface functions are actually implementation details (un/marshalling functions, client id generation etc. )

Yep I agree with that. It's indeed not good to have an interface that includes a number of functions that concern implementation details.

> I was investigating an [alternative approach](https://github.com/cosmos/ibc-go/pull/3214) which would allow custom logic to be executed at different client events (creation/upgrading/updating) through the use of hooks. I believe this should be enough to enable any sort of custom logic that is required while also making sure that the client keeper logic is guaranteed to execute as normal.

Thanks for your investigation and work. I believe adding relevant hooks directly to the client keeper suffices for our use cases and is much cleaner than the interface approach.

One thing is that our use case needs to distinguish between a header in the canonical chain and in the fork (i.e., has a valid quorum certificate but has non-monotonic timestamps or conflicting header). What we did is to provide the hook with a boolean value which indicates whether the header is canonical or fork (see [here](https://github.com/babylonchain/ibc-go/blob/generalize-babylon-callback/modules/core/02-client/keeper/extended_keeper.go#L82) and [here](https://github.com/babylonchain/ibc-go/blob/generalize-babylon-callback/modules/core/02-client/keeper/extended_keeper.go#L110)). This is difficult to be done in an external module subscribing to client keeper hooks since the external module does not have access to `clientState`s. Is it possible for you to implement something similar? Thanks!

crodriguezvega commented:
Closing for now, since the use case can be accomplished with SDK post handlers.
